fs_test.go

```diff
@@ -0,0 +1,303 @@
+package gofs_test
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/NeowayLabs/gofs"
+)
+
+type fsBuilder func(t *testing.T) gofs.FS
+
+func testFS(t *testing.T, newfs fsBuilder) {
```

Comment:
interesting the way tests are reused =)

Reply:
Besides reuse I tried to capture parts of the protocol that Go interfaces cannot express but are part of the interface protocol. For example, io.Reader will return EOF on the final Read call, but this kind of thing can't be expressed on the interface definition, neither can it be validated by the compiler, but it is an essential part of any implementation of an io.Reader.

I had implemented similar things on the past on C++ using inheritance, it was much more chaotic and harder to understand, God bless Go =)